### PR TITLE
[WIP] Initial implementation of single dispatch

### PR DESCRIPTION
**Pull request by certik**
This PR is to figure out the right approach (and for benchmarking various designs, as different commits). Once we settle on a design, we'll create a nice new polished PR and close this one.

### CODE REVIEW

File: src/add.cpp [hunk 1]
```diff
@@ -330,8 +330,9 @@ RCP<const Basic> sub(const RCP<const Basic> &a, const RCP<const Basic> &b)
     return add(a, mul(minus_one, b));
 }
 
-RCP<const Basic> add_expand(const RCP<const Add> &self)
+RCP<const Basic> add_expand(const RCP<const Basic> &self_)
 {
+    const RCP<const Add> self = rcp_static_cast<const Add>(self_);
```
Comment (certik) [inline src/add.cpp]: I don't like that the casting happens in these routines. It should happen before `add_expand` is called, but I can't figure out how to do that fast yet.

File: src/basic.cpp [hunk 2]
```diff
@@ -34,11 +34,18 @@ std::string Basic::__str__() const
...
-    if (is_a_Number(*self)) return self;
-    if (is_a<Add>(*self)) return add_expand(rcp_static_cast<const Add>(self));
-    if (is_a<Mul>(*self)) return mul_expand(rcp_static_cast<const Mul>(self));
-    if (is_a<Pow>(*self)) return pow_expand(rcp_static_cast<const Pow>(self));
+    typedef RCP<const Basic> (*fn)(const RCP<const Basic> &);
+    std::vector<fn> table;
+    table.assign(100, NULL);
+    table[ADD] = add_expand;
+    table[MUL] = mul_expand;
+    table[POW] = pow_expand;
```
Comment (certik) [inline src/basic.cpp]: This needs to happen only once, not everytime the `expand` is called.

File: src/add.cpp (hunk 1 above)
Comment (isuruf) [inline src/add.cpp]: This is an alternative, but I don't like the declaration of new functions. (eg: `rcp_static_cast_Add`)
```C++
RCP<const Mul> rcp_static_cast_Mul(const RCP<const Basic> &self) {
    return rcp_static_cast<const Mul>(self);
}

RCP<const Pow> rcp_static_cast_Pow(const RCP<const Basic> &self) {
    return rcp_static_cast<const Pow>(self);
}

RCP<const Add> rcp_static_cast_Add(const RCP<const Basic> &self) {
    return rcp_static_cast<const Add>(self);
}

RCP<const Basic> expand(const RCP<const Basic> &self)
{
    typedef std::function<RCP<const Basic>(const RCP<const Basic>)> fn;
    using std::placeholders::_1;
    std::vector<fn> table;
    table.assign(100, NULL);
    table[ADD] = std::bind(add_expand, std::bind(rcp_static_cast_Add, _1));
    table[MUL] = std::bind(mul_expand, std::bind(rcp_static_cast_Mul, _1));
    table[POW] = std::bind(pow_expand, std::bind(rcp_static_cast_Pow, _1));
    fn f = table[self->get_type_code()];
    if (f == NULL) {
        return self;
    } else {
        return f(self);
    }
    return self;
}
```

Comment (certik) [inline src/add.cpp]: Nice! I couldn't figure out how to do this. I think this is an improvement.
The declared functions can be reused and automatically declared using
macros.

I don't know if there is even better way.

Comment (shivamvats) [inline src/add.cpp]: This is good indeed. I must admit, my C++ needs more work.

Comment (certik) [inline src/add.cpp]: @isuruf I pushed your implementation into my PR. But I think there is significant overhead to `std::function` and `std::bind`, so we might need to figure out an alternative. But it seems to work.

**Comment by isuruf** [comment]
Build is failing because `#include <functional>` is missing.

**Comment by certik** [comment]
@isuruf I figured out how to get rid of the extra functions. Now it is general, but slow. I wonder if using perhaps Lambda functions could avoid using `std::bind` and `std::function`.

**Comment by isuruf** [comment]
Awesome. Now I understand why `rcp_static_cast` couldn't be used directly. 
This PR is not slower. It is faster than the master once the table is created outside the expand function and reused.

**Comment by certik** [comment]
Very good. Which benchmark did you use?

Master was slow, so this might be an improvement in this particular case, but ultimately we need to compare to `eval_double.cpp` and see how it compares to the visitor pattern (which I think we implemented very efficiently in csympy).

**Comment by isuruf** [comment]
I used `expand1` with this patch 
```diff
diff --git a/benchmarks/expand1.cpp b/benchmarks/expand1.cpp
index 8844dde..75b86f5 100644
--- a/benchmarks/expand1.cpp
+++ b/benchmarks/expand1.cpp
@@ -33,11 +33,12 @@ int main(int argc, char* argv[])
     RCP<const Basic> y = rcp(new Symbol("y"));
     RCP<const Basic> z = rcp(new Symbol("z"));
     RCP<const Basic> w = rcp(new Symbol("w"));
-    RCP<const Basic> i60 = rcp(new Integer(60));
+    RCP<const Basic> i2 = rcp(new Integer(2));
 
-    RCP<const Basic> e, r;
+    RCP<const Basic> e = x, r;
 
-    e = pow(add(add(add(x, y), z), w), i60);
+    for(int i = 0; i < 5; i++)
+        e = pow(add(mul(add(e, y), z), w), i2);
 
     std::cout << "Expanding: " << *e << std::endl;
```

**Comment by certik** [comment]
Nice. What timings did you get?

**Comment by isuruf** [comment]
15756ms(PR) vs 16250ms (master)

**Comment by certik** [comment]
Nice. Let's use this as a benchmark for this case.

I am trying to see if we can replace this:
```
+    table[ADD] = std::bind(add_expand, std::bind(rcp_static_cast_T<Add>, _1));
```
with lambda functions. The `std::function` documentation says that you should be able to assign lambda functions to it. I think that might be faster than `std::bind`.

**Comment by certik** [comment]
@isuruf would you mind benchmarking 426bdd000c947e70adac396dd99365e47edaf5de? I think it might be faster than `std::bind`.

**Comment by certik** [comment]
Also, could you separately benchmark c0e5f02ff21e27964a61fd3f474c9f37f466a636? So that we know if `std::function` has any overhead.

**Comment by certik** [comment]
Cool, I am done with the simplifications. It looks very nice and simple now. Let's see the benchmarks.

**Comment by isuruf** [comment]
For 426bdd000c947e70adac396dd99365e47edaf5de, I am getting 17600ms
For c0e5f02ff21e27964a61fd3f474c9f37f466a636, I am getting 18274ms

Can somebody else benchmark this and verify?

**Comment by certik** [comment]
So it got slower? That's unexpected. I'll try to benchmark later today.

**Comment by certik** [comment]
Indeed, I am getting similar benchmarks.

Hash | Time [ms]
------- | -------------
5db35dc | 20293
8bbfc2c | 19995
426bdd0 | 19984
c0e5f02 | 20432
131e2ab | 20423

We'll have to run these on more platforms, as my laptop is not very representative.

**Comment by certik** [comment]
I did some reading and it looks like `std::function` is actually very fast (with good compiler and all optimizations on), and it hides dealing with raw pointers, which we want anyway, so it looks like we should be using it.

Here is a benchmark: http://www.gockelhut.com/c++/files/lambda_vs_bind.cpp

@isuruf would you mind running it? Make sure you compiler it as follows:
```
$ g++ -std=c++11 -fPIC -O3 -march=native -ffast-math -funroll-loops lambda_vs_bind.cpp 
$ ./a.out
Accumulate (lambda)            	423390646
Accumulate (bind)              	3116846169
Accumulate (bound lambda)      	2704685011
```
(It also benchmarks Boost, but I don't have it installed.)

It shows that using lambdas + `std::function` is the fastest, faster than using raw pointer, which seems to be consistent with our benchmarks.

Can you double check, that `std::bind` is indeed slower than using lambdas, as long as you use `std::function`?

**Comment by certik** [comment]
I've implemented the same approach for `eval_double`. I think it looks a lot simpler than the visitor pattern. Now we need to benchmark it.

**Comment by certik** [comment]
My quick benchmarks show that it is maybe 30% slower than the visitor pattern. We need to create good benchmarks for the `eval_double` first.

**Comment by isuruf** [comment]
I got this benchmark
``` 
Accumulate (lambda)             220710975
Accumulate (bind)               2589433381
Accumulate (bound lambda)       2549014047
```

**Comment by certik** [comment]
Which compiler and version do you use?

**Comment by isuruf** [comment]
gcc 4.8.2

**Comment by certik** [comment]
The same as me. In any case, we should figure out a way to be as fast as visitor pattern for the `eval_double`.

**Comment by isuruf** [comment]
For `eval_double` what benchmark did you use?

**Comment by certik** [comment]
I quickly created one. We should create a larger expression and put it into
our benchmarks.

**Comment by certik** [comment]
But the problem with a large expression is that double does not have enough precision to evaluate it, so a better benchmark is to use a smaller expression (perhaps 10 or so operations) in terms of "x" and "y" as a 2D function and then evaluate for "x" and "y" from some region, e.g. for a 3D plot. That way we can make it arbitrary long. Since subs would have to be used, perhaps a good benchmark is to just assemble the expression using `mul`, `add`, ..., for each numerical value "x" and "y", avoiding subs. And just benchmark the whole thing.

**Comment by certik** [comment]
@isuruf I created `eval_double1` benchmark and if I run it with 5021b378c2e4134d11a473bd8c6ae15a3fa8d166, i.e. the single dispatch via lambda + `std::function` I get:

6463ms

While if I use the visitor pattern, I get:

4493ms

So the single dispatch is about 40% slower currently. It would be actually also very interesting to do this benchmark using just virtual functions, that should be even faster than the visitor pattern.

Let's see if we can make the single dispatch faster. If we can get to perhaps 10% (i.e. ~5000ms) difference to visitor pattern, then I think it could be usable --- we can always resolve to visitor pattern for extra efficiency if needed. Still, I would really like to be as fast as visitor pattern.

**Comment by certik** [comment]
Btw, on that test on my other machine (gcc 4.9.2), I get:
```
Accumulate (lambda)            	9918
Accumulate (bind)              	2100936483
Accumulate (bound lambda)      	1809219252
```
The first line is incorrect I think, things got optimized out. When I change `upper_limit = 100ULL`, I get:
```
Accumulate (lambda)            	9359
Accumulate (bind)              	4051
Accumulate (bound lambda)      	3492
```
In any case, looks like bound lambda, i.e. `std::function` + lambda, is the fastest.

**Comment by certik** [comment]
Ha, with dbc749425203c67b32d4e85a30aa9051bdd710b8 I am getting:

4544ms

Initially I thought it is the if statement (that can slow things down in inner loops), but it is rather the difference between:
```c++
    fn f = table_eval_double[b.get_type_code()];
    return f(b);
```
and
```c++
    return table_eval_double[b.get_type_code()](b);
```
I would think that the compiler would generate an equivalent code, but I guess it doesn't, perhaps due to the extra instantiation of `fn` (`std::function`).

**Comment by certik** [comment]
@isuruf would you mind running the `eval_double1` benchmark on your computer? Make sure you do Release mode with default options (i.e. start from a clean repository so that old CMake options do not affect things) and run it couple times, e.g. I get (for b96ff34f5f290645387bf6431bea80ad283c1d63):
```
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4706ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4785ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4471ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4604ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4520ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
```
Then apply this patch:
```diff
diff --git a/src/eval_double.cpp b/src/eval_double.cpp
index b1b78ad..cc9fd08 100644
--- a/src/eval_double.cpp
+++ b/src/eval_double.cpp
@@ -235,11 +235,8 @@ std::vector<fn> table_eval_double = init_eval_double();
 
 double eval_double(const Basic &b)
 {
-/*
     EvalDoubleVisitor v;
     return v.apply(b);
-*/
-    return table_eval_double[b.get_type_code()](b);
 }
 
 } // CSymPy
```
Recompile and run it again:
```
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4805ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4749ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4761ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4623ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4754ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
```
Looks like our new approach is even a tiny bit faster (perhaps) than visitor pattern.

**Comment by certik** [comment]
Further possible speedup could be done using #387, but that's a whole separate issue.

**Comment by certik** [comment]
Let's merge #388 and #389 first, but I've also updated `expand()` in b2a6607c5bc1b92739561bcf3e2710ee70967837, I will make PRs out of it later. Benchmarks before:
```
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./expand1 
Expanding: (w + (y + (w + (y + (w + (y + (w + (y + (w + (y + x)*z)^2)*z)^2)*z)^2)*z)^2)*z)^2
13352ms
number of terms: 6382
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./expand1 
Expanding: (w + (y + (w + (y + (w + (y + (w + (y + (w + (y + x)*z)^2)*z)^2)*z)^2)*z)^2)*z)^2
13315ms
number of terms: 6382
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./expand1 
Expanding: (w + (y + (w + (y + (w + (y + (w + (y + (w + (y + x)*z)^2)*z)^2)*z)^2)*z)^2)*z)^2
13333ms
number of terms: 6382
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./expand1 
Expanding: (w + (y + (w + (y + (w + (y + (w + (y + (w + (y + x)*z)^2)*z)^2)*z)^2)*z)^2)*z)^2
13364ms
number of terms: 6382
```
and after
```
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./expand1 
Expanding: (w + (y + (w + (y + (w + (y + (w + (y + (w + (y + x)*z)^2)*z)^2)*z)^2)*z)^2)*z)^2
13197ms
number of terms: 6382
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./expand1 
Expanding: (w + (y + (w + (y + (w + (y + (w + (y + (w + (y + x)*z)^2)*z)^2)*z)^2)*z)^2)*z)^2
13129ms
number of terms: 6382
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./expand1 
Expanding: (w + (y + (w + (y + (w + (y + (w + (y + (w + (y + x)*z)^2)*z)^2)*z)^2)*z)^2)*z)^2
13130ms
number of terms: 6382
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./expand1 
Expanding: (w + (y + (w + (y + (w + (y + (w + (y + (w + (y + x)*z)^2)*z)^2)*z)^2)*z)^2)*z)^2
13109ms
number of terms: 6382
```
So things got a bit faster.

**Comment by certik** [comment]
Finally, I also implemented virtual functions in c32c812dd47bfc0bebddd6d3d3ca7adc2310f35f. Test it like this:
```diff
diff --git a/benchmarks/eval_double1.cpp b/benchmarks/eval_double1.cpp
index bed80ec..847ac0a 100644
--- a/benchmarks/eval_double1.cpp
+++ b/benchmarks/eval_double1.cpp
@@ -37,7 +37,7 @@ int main(int argc, char* argv[])
     RCP<const Basic> e = sin(integer(1));
     double r, r_exact;
 
-    for(int i = 0; i < 10000; i++)
+    for(int i = 0; i < 9000; i++)
         e = pow(add(mul(add(e, pow(integer(2), integer(-3))), integer(3)), integer(1)), div(integer(2), integer(3)));
 
 //  Too long:
diff --git a/src/eval_double.cpp b/src/eval_double.cpp
index a36a139..c0ba864 100644
--- a/src/eval_double.cpp
+++ b/src/eval_double.cpp
@@ -239,7 +239,8 @@ double eval_double(const Basic &b)
     EvalDoubleVisitor v;
     return v.apply(b);
 */
-    return table_eval_double[b.get_type_code()](b);
+//    return table_eval_double[b.get_type_code()](b);
+    return b.eval_double();
 }
 
 } // CSymPy
```
Unfortunately one has to lower the nesting from 10000 to 9000, otherwise I get a stack overflow. So with 9000, before (with single dispatch):
```
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
3965ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4156ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
3938ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4172ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
3973ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
```
and after (virtual functions):
```
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4483ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4774ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4556ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
certik@redhawk:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
4624ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
```
So it seems our single dispatch is actually faster than virtual functions! That's unexpected. I implemented the virtual functions in headers, so that can't be slower than putting them in `cpp` files.

**Comment by certik** [comment]
@isuruf, @thilinarmtb if you can also run the benchmarks for virtual functions, visitor pattern and single dispatch on your computer (essentially just use e6c9eb251c814167825d831fc8f7638744dbe83a, and comment out the proper line in `eval_double()` in `src/eval_double.cpp`) and post it, that would be awesome (post compiler version as well).

I'll do the same on couple different platforms as well. Let's make sure that the single dispatch indeed is the fastest of all the alternatives. I hope it is, but let's see.

**Comment by certik** [comment]
On my laptop, I get (with gcc 4.8.2):

Virtual functions:
```
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
8881ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
9034ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
8999ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
8808ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
8921ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
```
Visitor Pattern:
```
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
9380ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
9329ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
9573ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
10164ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
9223ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
```
And single dispatch:
```
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
8991ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
9013ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
9164ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
9343ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
ondrej@eagle:~/repos/csympy/benchmarks(single_dispatch)$ ./eval_double1 
9654ms
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
```

My laptop is not very reliable to run benchmarks (the CPU speed and the computer performance is not constant), but it seems to show that visitor pattern is slower than virtual functions (as it should) and that single dispatch seems to be a little bit slower than virtual functions, but faster than visitor pattern.

**Comment by isuruf** [comment]
This is what I get.
```
Visitor
6356ms
6493ms
6369ms
6574ms
6483ms

Single-Dispatch
7493ms
6835ms
6529ms
7496ms
6479ms

Virtual Function
6262ms
6276ms
6259ms
6337ms
6500ms
```

Result is always
```
r (double) = 9.85647
r (exact)  = 9.85647
error      = 5.32907e-15
```

**Comment by isuruf** [comment]
All of the changes have been merged into master in several PRs

**Pull request closed without merging**